staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -444,6 +482,38 @@ func (r *crdHandler) serveScale(w http.ResponseWriter, req *http.Request, reques
 	}
 }
 
+// isAllowedToServe checks if the handler is allowed to serve without waiting for latest storageversion udpate.
+func isAllowedToServe(crdHandler *crdHandler, crdInfo *crdInfo, crd *apiextensionsv1.CustomResourceDefinition, requestInfo *apirequest.RequestInfo, w http.ResponseWriter, req *http.Request) error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		return nil
+	}
+
+	if alwaysServe := crdInfo.alwaysServe.Load(); alwaysServe {
+		klog.V(4).Infof("handler is allowed to serve always for crd:%s, proceeding to serve request", crd.Name)
+		return nil
+	}
+
+	ok, err := crdHandler.storageVersionManager.StorageVersionFoundInCacheFor(crd)
```

Comment:
Given an arbitrarily latent list/watch for StorageVersions, how does this prevent serving in a case of

1. steady state CRD storing v1
2. steady state StorageVersion encoding as v1
3. update CRD to storing v2
4. update StorageVersion encoding as v2, BUT watch event is delayed by five minutes
5. update CRD to storing v1
6. CR create arrives and is stored as v1, but StorageVersion indicates v2.

Reply:
Good catch. I think this is not handled today.

Reply:
Here's my understanding of this case:

1. SV was written to v1 correctly, SV cache reflects that.
2. CRD updated to v2, SV updated to v2, but SV cache is still at v1.
3. CRD updated to v1 again, CR request received - finds CRD SV to match SV cache - stores in v1.
4. the SV worker after 3. will find that CRD cache is at v1, SV cache is at v1, so will not process the v1 update again. Thus SV will be stuck at v2.

So to confirm, is the concern that the storageversion API will show v2 which is wrong since it should show v1?

Reply:
> So to confirm, is the concern that
> 
>     1. the storageversion API will show v2 which is wrong since it should show v1?
>     2. CR was allowed to be written in v1 before the SV was updated to v1?

Well a fix could come in either direction, but I guess the more specific problem is CR was allowed to be written in v1 before the storageversion was updated to v1.

Reply:
Had a proposal to address this issue (https://github.com/kubernetes/kubernetes/pull/124543#pullrequestreview-2025938219, [doc](https://docs.google.com/document/d/1-wiczjRXdWcJJPoBsxJkpvTkn3iGbJsSeAgzW4UsapM/edit?disco=AAABLNFyc3o)). @deads2k Interested to hear your opinion on it.